File: utils/text_similarity.go
package utils

import (
	"strings"
	"unicode"
)

// NormalizeText lowercases, trims, and collapses whitespace for text comparison.
func NormalizeText(s string) string {
	s = strings.TrimSpace(s)
	var b strings.Builder
	b.Grow(len(s))
	prevSpace := false
	for _, r := range s {
		if unicode.IsSpace(r) {
			if !prevSpace {
				b.WriteRune(' ')
				prevSpace = true
			}
			continue
		}
		b.WriteRune(unicode.ToLower(r))
		prevSpace = false
	}
	return strings.TrimSpace(b.String())
}

// TextSimilarity returns a 0–1 score for how alike two text blobs are.
// Uses exact match, substring overlap, and word Jaccard — catches copy-paste
// duplicates even when embeddings are missing or stale.
func TextSimilarity(a, b string) float32 {
	a = NormalizeText(a)
	b = NormalizeText(b)
	if a == "" || b == "" {
		return 0
	}
	if a == b {
		return 1
	}

	// One string fully contained in the other
	if strings.Contains(a, b) || strings.Contains(b, a) {
		shorter, longer := len(a), len(b)
		if shorter > longer {
			shorter, longer = longer, shorter
		}
		ratio := float32(shorter) / float32(longer)
		if ratio > 0.85 {
			return ratio
		}
	}

	wordsA := strings.Fields(a)
	wordsB := strings.Fields(b)
	if len(wordsA) == 0 || len(wordsB) == 0 {
		return 0
	}

	setA := make(map[string]struct{}, len(wordsA))
	for _, w := range wordsA {
		setA[w] = struct{}{}
	}

	intersection := 0
	setB := make(map[string]struct{}, len(wordsB))
	for _, w := range wordsB {
		if _, seen := setB[w]; seen {
			continue
		}
		setB[w] = struct{}{}
		if _, ok := setA[w]; ok {
			intersection++
		}
	}

	union := len(setA) + len(setB) - intersection
	if union == 0 {
		return 0
	}
	return float32(intersection) / float32(union)
}

// CombinedSimilarity returns the best of vector and text similarity.
func CombinedSimilarity(query string, captureTitle, captureBody string, captureTags []string, queryVector, captureVector []float32) float32 {
	captureText := CaptureSearchText(captureTitle, captureBody, captureTags)
	textScore := TextSimilarity(query, captureText)

	var vectorScore float32
	if len(queryVector) > 0 && len(captureVector) > 0 {
		vectorScore = CosineSimilarity(queryVector, captureVector)
	}

	if textScore > vectorScore {
		return textScore
	}
	return vectorScore
}

File: frontend/src/lib/textSimilarity.js
/**
 * Text similarity for duplicate detection — mirrors utils/text_similarity.go.
 * Catches copy-paste duplicates when fingerprints are missing (pre-v2 notes).
 */

function normalizeText(s) {
  const trimmed = (s ?? '').trim().toLowerCase();
  const parts = trimmed.split(/\s+/).filter(Boolean);
  return parts.join(' ');
}

/** @returns {number} 0–1 score */
export function textSimilarity(a, b) {
  const na = normalizeText(a);
  const nb = normalizeText(b);
  if (!na || !nb) return 0;
  if (na === nb) return 1;

  if (na.includes(nb) || nb.includes(na)) {
    const shorter = na.length <= nb.length ? na.length : nb.length;
    const longer = na.length > nb.length ? na.length : nb.length;
    const ratio = shorter / longer;
    if (ratio > 0.85) return ratio;
  }

  const wordsA = na.split(' ');
  const wordsB = nb.split(' ');
  if (!wordsA.length || !wordsB.length) return 0;

  const setA = new Set(wordsA);
  const setB = new Set(wordsB);
  let intersection = 0;
  for (const w of setB) {
    if (setA.has(w)) intersection += 1;
  }
  const union = setA.size + setB.size - intersection;
  if (union === 0) return 0;
  return intersection / union;
}

/** Build comparable text blob from a decrypted capture (title + tags + body). */
export function captureComparableText(capture) {
  const parts = [];
  const tags = (capture?.tags ?? []).join(' ');
  if (tags) parts.push(tags);
  const title = (capture?.title ?? '').trim();
  if (title) parts.push(title);
  const body = (capture?.body ?? '').trim();
  if (body) parts.push(body);
  return parts.join('\n');
}
